core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -867,22 +867,40 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
   ): ZIO[R1, E2, C] = {
     def arbiter[E0, E1, A, B](
       f: (Exit[E0, A], Fiber[E1, B]) => ZIO[R1, E2, C],
+      winner: Fiber[E0, A],
       loser: Fiber[E1, B],
-      race: Ref[Int],
+      waitForLeft: Boolean,
+      race: Ref[Boolean],
+      inherit: Ref[Option[Boolean]],
       done: Promise[E2, C]
-    )(res: Exit[E0, A]): ZIO[R1, Nothing, _] =
-      ZIO.flatten(race.modify((c: Int) => (if (c > 0) ZIO.unit else f(res, loser).to(done).unit) -> (c + 1)))
+    )(res: Exit[E0, A]): ZIO[R1, Nothing, _] = {
+
+      val handleRes =
+        (f(res, loser) <* {
+          ZIO.when(res.succeeded)(winner.inheritFiberRefs) *> inherit.set(Some(waitForLeft))
+        }).to(done)
+
+      ZIO.flatten(race.modify(b => (if (b) ZIO.unit else handleRes) -> true))
+    }
 
     for {
-      done <- Promise.make[E2, C]
-      race <- Ref.make[Int](0)
+      done    <- Promise.make[E2, C]
+      race    <- Ref.make[Boolean](false)
+      inherit <- Ref.make(None: Option[Boolean])
       c <- ZIO.uninterruptibleMask { restore =>
             for {
-              left  <- ZIO.interruptible(self).fork
-              right <- ZIO.interruptible(that).fork
-              _     <- left.await.flatMap(arbiter(leftDone, right, race, done)).fork
-              _     <- right.await.flatMap(arbiter(rightDone, left, race, done)).fork
-              c     <- restore(done.await).onInterrupt(left.interrupt *> right.interrupt)
+              left   <- ZIO.interruptible(self).fork
+              right  <- ZIO.interruptible(that).fork
+              left2  <- left.await.flatMap(arbiter(leftDone, left, right, true, race, inherit, done)).fork
+              right2 <- right.await.flatMap(arbiter(rightDone, right, left, false, race, inherit, done)).fork
+
+              inheritFiberRefs = inherit.get.flatMap {
+                case None        => ZIO.unit
+                case Some(true)  => left2.inheritFiberRefs
```

Comment:
There is a more direct way to encode this: use `Ref[Task[Unit]]`. Then you can start that `Ref` off with `ZIO.unit`, and then you can either stick `left2.inheritFiberRefs` into that, or `right2.inheritFiberRefs` into it, depending on the race outcome.

This will simplify the code a _lot_.